service/src/main/java/org/dice_research/fc/sparql/query/QueryEngineCustomHTTP.java

```diff
@@ -0,0 +1,225 @@
+    private long timeout=15000;
+
+    public QueryEngineCustomHTTP(Query query, String service) {
+        this.query = query;
+        this.service = service;
+    }
+
+    @Override
+    public void setInitialBinding(QuerySolution binding) {
+
+    }
+
+    @Override
+    public Dataset getDataset() {
+        return null;
+    }
+
+    @Override
+    public Context getContext() {
+        return null;
+    }
+
+    @Override
+    public Query getQuery() {
+        return this.query;
+    }
+
+    @Override
+    public ResultSet execSelect() {
+        String result = createRequest(query.toString(), timeout, service);
+        ResultSetFactory fac = new ResultSetFactory();
+
+        // the result is not a valid XML then replace with an empty XML
+        if(result.length()<10) {
+            result = emptyXML();
+        }
+
+        ResultSet resultSet = fac.fromXML(result);
+
+        return resultSet;
+    }
+
+    private String emptyXML() {
+        return "<sparql xmlns=\"http://www.w3.org/2005/sparql-results#\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"http://www.w3.org/2001/sw/DataAccess/rf1/result2.xsd\"><head></head><results distinct=\"false\" ordered=\"true\"></results></sparql>>";
+    }
+
+    private static String read(InputStream content) {
+        ByteSource byteSource = new ByteSource() {
+            @Override
+            public InputStream openStream() throws IOException {
+                return content;
+            }
+        };
+
+        try {
+            return byteSource.asCharSource(Charsets.UTF_8).read();
+        } catch (IOException e) {
+            LOGGER.error("Could not read stream due to ",e);
+        }
+        return "";
+
+    }
+
+    protected static String createRequest(String sparqlQuery, Long to,String service){
+        List<String> ret =  new ArrayList<String>();
+        int code=0;
+        String test;
+        String actualContentType="";
+        try {
+            int timeout = to.intValue();
+            RequestConfig config = RequestConfig.custom()
+                    .setConnectTimeout(timeout)
+                    .setConnectionRequestTimeout(timeout)
+                    .setSocketTimeout(timeout).build();
+            CloseableHttpClient client = HttpClientBuilder.create().setDefaultRequestConfig(config).build();
+            HttpGet get = new HttpGet(service+"?query="+ URLEncoder.encode(sparqlQuery, "UTF-8"));
+            get.addHeader(HttpHeaders.ACCEPT, "application/sparql-results+xml");
+            HttpResponse resp = client.execute(get);
+            test = read(resp.getEntity().getContent());
+            InputStream is = new ByteArrayInputStream(test.getBytes(StandardCharsets.UTF_8));
+            String result = IOUtils.toString(is, StandardCharsets.UTF_8);
+            LOGGER.info(result);
+            return result;
+        }catch(Exception e){
+            LOGGER.error("Could not execute request due to ",e);
```

Comment:
This behavior of the class will lead to the nearly silent ignoring of errors. Yes, the error is printed but the program that called the class has no chance to be informed that something went wrong. Instead, an empty SELECT is created which will lead to an empty (but valid) result. Hence, the program will think that everything worked fine but the dataset does not contain the piece of information the select query tried to retrieve.

Please check what kind of exception the execSelect() method is allowed to throw. I am pretty sure that there is something that we can use :thinking:

Reply:
the timeout throws an exception, then I handle timeout separately and for other exceptions used this
```
throw new RuntimeException("There is an error while running the query",e);
```